roofit/batchcompute/src/Initialisation.cxx

```diff
@@ -1,75 +1,72 @@
-#include "RooBatchCompute.h"
+#include "rbc.h"
 
 #include "TEnv.h"
 #include "TSystem.h"
+#include "TError.h"
 
-#include <iostream>
 #include <string>
 #include <exception>
 
-
-// First initialisation of the pointer. When implementations of the batch compute library are loaded,
-// they will overwrite the pointer.
-RooBatchCompute::RooBatchComputeInterface* RooBatchCompute::dispatch=nullptr;
+// First initialisation of the pointers. When implementations of the batch compute library
+// are loaded, they will overwrite the pointers.
+RooBatchCompute::RbcInterface *RooBatchCompute::dispatchCPU = nullptr;
+RooBatchCompute::RbcInterface *RooBatchCompute::dispatchCUDA = nullptr;
 
 namespace {
+  
+/// Dynamically load a library and throw exception in case of failure
+void loadWithErrorChecking(const std::string& libName)
+{
+  const auto returnValue = gSystem->Load(libName.c_str());
+  if (returnValue == -1 || returnValue == -2)
+    throw std::runtime_error("RooFit was unable to load its computation library " + libName);
+  else if (returnValue == 1) // Library should not have been loaded before we tried to do it.
+    throw std::logic_error("RooFit computation library " + libName + " was loaded before RooFit initialisation began.");
+}
 
-////////////////////////////////////////////////////////////////////////////////////////////////////////////////
-/// Inspect cpu capabilities, and load the optimal library for RooFit computations.
-void loadComputeLibrary() {
-
-  std::string libName="libRooBatchCompute_GENERIC";
-
+/// Inspect hardware capabilities, and load the optimal library for RooFit computations.
+void loadComputeLibrary()
+{
+  const std::string userChoice = gEnv->GetValue("RooFit.BatchCompute","auto");
 #ifdef R__RF_ARCHITECTURE_SPECIFIC_LIBS
-  
-  __builtin_cpu_init();
-  if (gEnv->GetValue("RooFit.LoadOptimisedComputationLibrary", 1) == 0) {
-    if (gDebug>0) {
-      std::cout << "In roofitcore/InitUtils.cxx:loadComputeLibrary(): RooFit.LoadOptimisedComputationLibrary is set to 0, using generic RooBatchCompute library." << std::endl;
-    }
-  }
-  
-  #if __GNUC__ > 5 || defined(__clang__)
-  //skylake-avx512 support
-  else if (__builtin_cpu_supports("avx512cd") && __builtin_cpu_supports("avx512vl") && __builtin_cpu_supports("avx512bw") && __builtin_cpu_supports("avx512dq"))  {
-    libName = "libRooBatchCompute_AVX512";
-  }
-  #endif 
-  
-  else if (__builtin_cpu_supports("avx2")) {
-    libName = "libRooBatchCompute_AVX2";
-  } else if (__builtin_cpu_supports("avx")) {
-    libName = "libRooBatchCompute_AVX";
-  } else if (__builtin_cpu_supports("sse4.1")) {
-    libName = "libRooBatchCompute_SSE4.1";
-  }
+  #ifdef R__HAS_CUDA
+    TString libcudart{"libcudart"};
+    if (gSystem->FindDynamicLibrary(libcudart, true)) gSystem->Load("libRooBatchCompute_CUDA");
```

Comment:
This doesn't work because `FindDynamicLibrary` takes a reference, not `const` reference:
```
cannot bind non-const lvalue reference of type ‘TString&’ to an rvalue of type ‘TString’
```